File: routeMatching.R
## R-Solution to process trippoints, route trippoints and plot 
## the route with road segments on map
## Author: Amir Abu Jandal

# ==> getSampleDF(): Function to get sample data in a dataframe
getSampleDF = function(file1, file2){
  
  #Reading the trippoints sample data
  sample_df <-read.csv(file1, header=FALSE, stringsAsFactors=TRUE,sep="\t",quote="")
  
  #Reading column names of the trippoints data
  names(sample_df) = read.table(file2, header = FALSE)[[1]]
  names(sample_df) = gsub(",", "", names(sample_df))
  
  return(sample_df)
}# END of getSampleDF()

# ==> getDatabaseConnection(): R connectivity with PostgreSQL
getDatabaseConnection = function(dbName, hostName, portNumber, userName, password){
  drv <- dbDriver("PostgreSQL")
  con <- dbConnect(drv, dbname=dbName, host=hostName, port=portNumber, user=userName, password=password)
  
  return(con)
}# END of getDatabaseConnection()

# ==> nearestNeighbourWay(): Get nearest way from PostGIS using the getBoundingBox() function
nearestNeighbourWay = function(tripPoint, dbConnection){
  lat = tripPoint$latitude
  lon = tripPoint$longitude
  boundingBox <- getBoundingBox(lat, lon, 1.2)
  
  minLon = boundingBox[1]
  minLat = boundingBox[2] 
  maxLon = boundingBox[3]
  maxLat = boundingBox[4]

  spatialQuery = paste("with distance_query as(SELECT DISTINCT ON (way)
                       osm_id, name, ST_Distance(ST_Transform(ST_GeomFromText('POINT(",lon," ",lat,")',4326),3857), ST_Transform(way, 3857)) AS myLineDistance, ST_AsText(ST_Transform(way, 4326)) as way
                       FROM public.planet_osm_line 
                       WHERE public.planet_osm_line.way && ST_Transform(ST_MakeEnvelope(",minLon,",", minLat,",", maxLon,",", maxLat,", 4326), 3857)
                       AND ST_Distance(ST_Transform(ST_GeomFromText('POINT(",lon," ",lat,")',4326),3857), ST_Transform(way, 3857)) < 10.0
                       AND osm_id > 0)
                       SELECT *FROM distance_query 
                       ORDER BY myLineDistance LIMIT 3;")
  
  #way: Execute the query and get the results in a simple df with geometry as text $$$ PLEASE UPDATE THE ABOVE QUERY $$$
  way = dbGetQuery(conn = con, spatialQuery)
  
  return(way)  
}# END of nearestNeighbourWay()

# ==> getMatchingPointData(): Get the matching point to the GPS point on the road
getMatchingPointData = function(road, gpsLat, gpsLon){
  distances = vector()
  
  for(node in road){
    latlon = strsplit(node, " ")[[1]]
    lat = as.numeric(latlon[2])
    lon = as.numeric(latlon[1])
    
    #d = distm(c(lat, lon), c(gpsLat, gpsLon), fun = distHaversine)[1][1]
    d = distGeo(matrix(c(lat, lon), ncol = 2), matrix(c(gpsLat, gpsLon), ncol = 2))/1609.34
    distances = c(distances, d)
  }
  
  minimumNodeIndex = which(distances == min(distances))
  matchedNode = road[minimumNodeIndex]
  latlon = strsplit(matchedNode, " ")[[1]]
  matchedLat = as.numeric(latlon[2])
  matchedLon = as.numeric(latlon[1])
  
  matchedCoordinates = c(matchedLat, matchedLon)
  
  return(list(matchedCoordinates=matchedCoordinates, minimumDistance=distances[minimumNodeIndex]))
}#END of getMatchingPointData()

# ==> viewMatchedPointsOnLeaflet(): View gps points and matched points on leaflet
viewMatchedPointsOnLeaflet = function(m, tripDF){
  
  gpxData = cbind(tripDF$latitude, tripDF$longitude)
  matchedData = cbind(tripDF$matchedLat, tripDF$matchedLon)
  
  for(k in 1:nrow(gpxData)){
    if(k == 1){
      m <- addCircleMarkers(m, color = "#31a354", radius = 4, lng = gpxData[k, 2], lat = gpxData[k, 1])   
      m <- addCircleMarkers(m, color = "red", radius = 5, lng = matchedData[k, 2], lat = matchedData[k, 1])
      
    }else if( k == nrow(gpxData)){
      m <- addCircleMarkers(m, color = "#31a354", radius = 4, lng = gpxData[k, 2], lat = gpxData[k, 1]) 
      m <- addCircleMarkers(m, color = "red", radius = 5, lng = matchedData[k, 2], lat = matchedData[k, 1])    
    }else{
      m <- addCircleMarkers(m, color = "#31a354", radius = 4, lng = gpxData[k, 2], lat = gpxData[k, 1]) 
      m <- addCircleMarkers(m, color = "red", radius = 5, lng = matchedData[k, 2], lat = matchedData[k, 1])
    }
  }
  
  return(m)
}#END of viewMatchedPointsOnLeaflet()

### SEGEMENTS can be EXTRACTED from THIS function!!!
# ==> viewRoadSnapOnLeaflet(): View road snap for a trip on leaflet
viewRoadSnapOnLeaflet = function(m, roadSnapDF, tripMatchDF){
  
  if(nrow(roadSnapDF) > 0){
    roadSegment = ""
    lonSegmentVector = vector()         # CAN BE USED TO COLLECT THE SEGMENTS
    latSegmentVector = vector()         # CAN BE USED TO COLLECT THE SEGMENTS
    startLon = vector()
    startLat = vector()
    endLon = vector()
    endLat = vector()
    
    for(w in 1:nrow(roadSnapDF)){
      lonVector = vector()
      latVector = vector()
      
      roadSegmentList = substr(roadSnapDF[w, ]$st_astext, 12, nchar(roadSnapDF[w, ]$st_astext)-1)
      roadSegment = strsplit(roadSegmentList, ",")[[1]]
      
      for(u in 1:length(roadSegment)){
        latlon = strsplit(roadSegment[u], " ")[[1]]
        lonVector = c(lonVector, as.numeric(latlon[1]))
        latVector = c(latVector, as.numeric(latlon[2]))
      }
      
      m <- addCircleMarkers(m, color = "blue", radius = 2, lng = lonVector, lat = latVector)
      m <- addPolylines(m, color = "black", lng = lonVector, lat = latVector)
      '
      if(roadSnapDF[w, ]$seq == 0){
        endLon = lonVector[1]
        endLat = latVector[1]
        
        if(length(startLon)>0 && length(startLon)>0){
          m <- addPolylines(m, color = "black", lng = lonSegmentVector, lat = latSegmentVector)
        }
        startLon = vector()
        startLat = vector()
      }else{
        startLon = lonVector[u]
        startLat = latVector[u]
      }
      
      lonSegmentVector = c(startLon, endLon)
      latSegmentVector = c(startLat, endLat)'
    }# END of outer FOR LOOP
  }
  
  return(m)
}#END of viewRoadSnapOnLeaflet()

# ==> getTripMatchDF()
getTripMatchDF = function(dfSubset, con){
  n = list("osm_id", "names", "mylinedistance", "way")
  way <- data.frame(matrix(nrow = 0, ncol = length(n)))
  names(way) <- n
  class(way$osm_id) <- "numeric"
  class(way$names) <- "character"
  class(way$mylinedistance) <- "numeric"
  class(way$way) <- "character"
  
  n = list("matchedLat", "matchedLon")
  matchedNode <- data.frame(matrix(nrow = 0, ncol = length(n)))
  names(matchedNode) <- n
  class(matchedNode$matchedLat) <- "numeric"
  class(matchedNode$matchedLon) <- "numeric"
  
  tripDF = merge(dfSubset[0, ], matchedNode[0, ])
  tripDF = merge(tripDF[0, ], way[0, ])
  class(tripDF$osm_id) <- "numeric"
  class(tripDF$names) <- "character"
  class(tripDF$mylinedistance) <- "numeric"
  class(tripDF$way) <- "character"
  
  # For Snapping the road segments
  roadSnapDF <- data.frame()
  
  # Tracking previous way's Name
  previousWayId = ""
  
  for(i in 1:nrow(dfSubset)){
    tripPoint <- dfSubset[i, ]
    way <- nearestNeighbourWay(tripPoint, con)
    
    if(i == 1){
      way <- way[1, ]
      roadString <- substr(way$way, 12, nchar(way$way)-1)
      road <- strsplit(roadString, ',')[[1]]
      matchedCoordinates <- getMatchingPointData(road, as.numeric(tripPoint$latitude), as.numeric(tripPoint$longitude))[[1]]
      matchedNode[i, ] <- c(matchedCoordinates[1], matchedCoordinates[2])
      
      tripDF[i, ] <- merge(merge(tripPoint, matchedNode), way)
    }else if(nrow(way) > 0){
      gpsLat = as.numeric(tripPoint$latitude)
      gpsLon = as.numeric(tripPoint$longitude)
      
      # w is the estimation value for TOP 3 ways. Where as, (w = ä.d + ß.t)
      distanceCost = vector(length = nrow(way))
      pathCost = vector(length = nrow(way))
      nearestWay = vector(length = nrow(way))
      wayEstimationValue = vector(length = nrow(way))
      shortestPathList = list(length = nrow(way))
      
      alpha = 0.35
      beta = 0.45
      gama = 1 - (alpha+beta)
      
      # Initial check to avoid caluclation of further ways
      #switchArray <- getSwitchArray(way)
      
      for(q in 1:nrow(way)){
        #if(q > 1 && switchArray[q-1] == 0){
        #  for(r in q:nrow(way)){
        #    wayEstimationValue[r] = 10000
        #  }
        #  break
        #}
        
        roadString <- substr(way[q, ]$way, 12, nchar(way[q, ]$way)-1)
        road <- strsplit(roadString, ',')[[1]]
        
        # Get matching point for GPS on the nearest road or way
        matchingPointData <- getMatchingPointData(road, gpsLat, gpsLon)
        matchedCoordinates <- matchingPointData[[1]]
        
        sourceLon <- format(round(matchedNode[i-1, ][2], 5), nsmall = 5)
        sourceLat <- format(round(matchedNode[i-1, ][1], 5), nsmall = 5)
        targetLon <- format(round(matchedCoordinates[2], 5), nsmall = 5)
        targetLat <- format(round(matchedCoordinates[1], 5), nsmall = 5)
        
        sourceNodeQuery <- paste("SELECT source FROM hh_2po_4pgr 
                                ORDER BY geom_way <-> ST_SetSRID(ST_Makepoint(", sourceLon, ", ", sourceLat,"), 4326) LIMIT 1;")
        targetNodeQuery <- paste("SELECT target FROM hh_2po_4pgr 
                                ORDER BY geom_way <-> ST_SetSRID(ST_Makepoint(", targetLon, ", ", targetLat,"), 4326) LIMIT 1;")
          
        sourceNode <- dbGetQuery(conn = con, sourceNodeQuery)[1, 1]
        targetNode <- dbGetQuery(conn = con, targetNodeQuery)[1, 1]
        
        shortestPathQuery <- paste("SELECT seq, id1 AS node, id2 AS edge, di.cost, ST_AsText(ST_Transform(geom_way, 4326)) 
                                   FROM pgr_dijkstra('SELECT id, source, target, st_length(geom_way) as cost FROM hh_2po_4pgr',",
                                   sourceNode, ", ", targetNode,", false, false ) as di JOIN hh_2po_4pgr ON di.id2 = hh_2po_4pgr.id")
        
        shortestPathList[[q]] <- dbGetQuery(conn = con, shortestPathQuery)

        matchingPointDistance <- matchingPointData[[2]]
        shortestPathCost <- sum(shortestPathList[[q]]$cost)
        nearestWayDistance <- way[q, ]$mylinedistance
        
        # FINALLY, the decision value can be stated as w = (ä.x) + (ß.y) + (g.z)
        distanceCost[q] <- matchingPointDistance
        pathCost[q] <- shortestPathCost
        nearestWay[q] <- nearestWayDistance
        
        if(is.na(way$name[q])){
          way$name[q] <- ""
        }
        
        if(previousWayId == way$name[q]){
          wayEstimationValue[q] <- -1
          break
        }else{
          wayEstimationValue[q] <- (alpha * matchingPointDistance) + (beta * shortestPathCost) + (gama * nearestWayDistance)   
        }
      }# END of q FOR LOOP
        
      # Getting the index of the way with least cost
      bestIndex <- which.min(wayEstimationValue)
      print(paste("Selected way = ", bestIndex))
      print(paste("Way Estimation values = ", wayEstimationValue))
      
      # Extracting the road segmetn
      roadString <- substr(way[bestIndex, ]$way, 12, nchar(way[bestIndex, ]$way)-1)
      road <- strsplit(roadString, ',')[[1]]
      
      # Getting the matched points on the road segment
      matchedCoordinates <- getMatchingPointData(road, gpsLat, gpsLon)[[1]]
      matchedNode[i, ] <- c(matchedCoordinates[1], matchedCoordinates[2])
      
      # Snapping the road segments 
      roadSnapDF <- rbind(roadSnapDF, shortestPathList[[bestIndex]])
      
      tripDF[i, ] <- merge(merge(tripPoint, matchedNode[i, ]), way[bestIndex, ])
    }else{
      matchedNode[1, ] <- c(0, 0)
      
      n <- list("osm_id", "names", "mylinedistance", "way")
      way <- data.frame(matrix(nrow = 0, ncol = length(n)))
      names(way) <- n
      class(way$osm_id) <- "numeric"
      class(way$names) <- "character"
      class(way$mylinedistance) <- "numeric"
      class(way$way) <- "character"
      way[1, ] <- c(0, "", 0, "")
      
      tripDF[i, ] <- merge(merge(tripPoint, matchedNode), way)
    }
    
    previousWayId = tripDF[i, ]$names
    if(is.na(previousWayId)){
      previousWayId <- ""
    }
    print(paste(100*i/nrow(dfSubset), "% has been completed!"))
  }# END of OUTER FOR LOOP
  
  # Plot the matched points and the GPS points
  m <- viewMatchedPointsOnLeaflet(m, tripDF)
  
  # Plotting the road segment
  m <- viewRoadSnapOnLeaflet(m, roadSnapDF, tripDF)
  
  return(list(trip=tripDF, map=m))
}# END of getTripMatchDF()

# ==> getBoundingBox(): Get a bounding box around lat/lon for some distance
getBoundingBox = function(radLat, radlon, distance){
  
  eRadius <- 6371.01
  
  pRadius = eRadius*cos(radLat)
  
  minLat <- radLat - distance/eRadius
  maxLat <- radLat + distance/eRadius
  minLon <- radlon - distance/pRadius
  maxLon <- radlon + distance/pRadius
  
  boundingBox <- c(minLon, minLat, maxLon, maxLat)
  
  return(boundingBox)
}# END of getBoundingBox()

# ==> getSwitchArray(): Initial check to avoid calculation of further ways and to mostly select first way
getSwitchArray = function(way){
  wayDistances <- way$mylinedistance
  returnArray <- vector(length = length(wayDistances)-1)
  
  if(length(wayDistances) > 1){
    for(j in 1:length(wayDistances)-1){
      diff <- abs(wayDistances[1] - wayDistances[j+1])
      
      print(paste("DIFFERENCE = ", diff))
      
      if(diff < 10){
        returnArray[j] = 1
      }else{
        returnArray[j] = 0
        break
      }
    }    
  }else{
    returnArray[1] = 0
  }
  print(paste("returnArray = ", returnArray))
  return(returnArray)
}# END of getSwitchArray()

# ==> mainFunction(): MAIN function
mainFunction = function(){
  library(sp)
  library(geosphere)
  library(leaflet)
  library(RPostgreSQL)
  
  # Reading the GPS data
  #tripPoints_file <- "/trippoint_sample2.tsv"
  tripPoints_file <- "/testAlgo.csv"
  header_file <- "/trippoint_headerInfo.txt"
  #sample_df <- getSampleDF(tripPoints_file, header_file)
  sample_df <- read.csv(tripPoints_file, header=FALSE, stringsAsFactors=FALSE,sep=",")
  names(sample_df) <- read.table(header_file, header = FALSE)[[1]]
  names(sample_df) <- gsub(",", "", names(sample_df))
  
  # Opening a connection to the database
  con <- getDatabaseConnection("trieste", "localhost", 5432, "postgres", "123456")
  
  # Initializing the Leaflet map variable
  options(viewer = NULL)
  m <- leaflet()
  m <- setView(m, lng = 13.7, lat = 45.6, zoom = 11)
  m <- addTiles(m)
  
  # Final dataframe with trippoints and their matching coordinates and roads
  tripMatchDF <- data.frame()
  
  # Unique trips
  trip_id <- unique(as.character(sample_df[,c("trip=trip_id")]))
  
  for(k in 1:length(trip_id)){
    startTime <- Sys.time()
    
    id <- trip_id[k]
    dfSubset <- sample_df[sample_df$'trip=trip_id' == id, ]
    
    # Ordering the trip from start to the end
    dfSubset <- dfSubset[order(dfSubset$corrected_timestamp), ]
    dfSubset <- dfSubset[order(dfSubset$point_timestamp), ]
    
    # Getting the road segments and matching points for the provided GPS trippoints
    returnedList <- getTripMatchDF(dfSubset, con)
    tripMatchDF <- rbind(tripMatchDF, returnedList$trip)
    m <- returnedList$map
    
    endTime <- Sys.time()
    executionTime <- endTime - startTime
    executionTime
  }# END of FOR LOOP
  
  View(tripMatchDF)
  
  ### UNCLEAR TRIPPOINTS
  # Reading the unclear trippoints
  tripPoints_file = "/unclearTrippoints.csv"
  unclear_df <-read.csv(tripPoints_file, header=TRUE, stringsAsFactors=FALSE, sep=",")
  
  # Initializing the Leaflet map variable
  options(viewer = NULL)
  m <- leaflet()
  m <- setView(m, lng = 13.7, lat = 45.6, zoom = 11)
  m <- addTiles(m)
  
  # Final dataframe with trippoints and their matching coordinates and roads
  tripMatchDF <- data.frame()
  
  for(i in 1:nrow(unclear_df)){
    # Getting the roads for provided GPS trippoints
    tripMatchDF <- rbind(tripMatchDF, getTripMatchDF(unclear_df[i, ], con))
  }
  
  
  #Disconnect from the server and the database
  dbDisconnect(con)
  dbUnloadDriver(drv)
  
}# END of mainFunction()

File: routeMatching.py
# -*- coding: utf-8 -*-
"""
Created on Thu Oct 27 10:33:22 2016

@author: Amir Abu Jandal
"""
import math
import folium
from sqlalchemy import create_engine
import pandas as pd
import numpy as np
from math import radians, cos, sin, asin, sqrt
import timeit
from IPython.display import HTML, display

#### mainMethod(): The main method to execute the whole algorithm
def mainMethod():    
    #==>[1]: Read the trip points data into a dataframe    
    tripPoints_file = "/testAlgo.csv"
    sample_df = pd.read_csv(tripPoints_file, sep=";")
    #header_file = "/trippoint_headerInfo.txt"    
    #sample_df.columns = open(header_file, 'r').readlines()[0].split(',')
    
    #==>[2]: Get a connection to PostgreSQL
    dbParameters = "dbname='trieste' user='postgres' host='localhost' password='123456'"
    con = getDBconnection(dbParameters)
    
    #==>[3]: Initialize a map and the final required dataframe
    map_osm = folium.Map(location=[45.6, 13.7], zoom_start= 11)
    map_osm.save(outfile='/osm.html')
    
    tripMatchDF = pd.DataFrame()
    
    #==>[4]: Loop through every trip and process it for matching and routing
    unique_ids = sample_df["trip_id"].unique()
    for i in range(len(unique_ids)):
        trip_id = unique_ids[i]
        dfSubset = sample_df[sample_df['trip_id'] == trip_id]
        
        #Getting appropriate index values for the dfSubset dataframe
        dfSubset.index = range(len(dfSubset.index.values))
        
        # Sorting the trip from start to the end
        dfSubset = dfSubset.sort(["corrected_timestamp"], ascending=[True])
        dfSubset = dfSubset.sort(["point_timestamp"], ascending=[True])
        
        # Getting the road segments and matching points for the provided GPS trippoints
        start_time = timeit.default_timer()
        returnedList = processTrip(dfSubset, con, map_osm)
        elapsed = timeit.default_timer() - start_time
        print("Time taken to process 1 trip is " + str(elapsed) + " seconds.")
        
        tripMatchDF.append(returnedList[0])
        map_osm = returnedList[1]
        map_osm.save(outfile='/osm_result.html')
    #END of FOR Loop
#END of mainMethod()  

#### getDBconnection(): get connection to spatial database
def getDBconnection(dbParameters):
    try:
        engine = create_engine('postgresql://postgres:123456@localhost:5432/trieste', encoding='latin1')
        con = engine.connect()
    except:
        print("Couldn't connect to the database!")
    
    return(con)
#END of getDBconnection()
    
#### processTrip(): process each trip to find matching points and road segments
def processTrip(dfSubset, con, map_osm):
    ways_df = pd.DataFrame(np.nan, index=range(len(dfSubset)), columns=list(["trip_id", "osm_id", "name", "mylinedistance", "way"]))
    ways_df.iloc[0] = [0, 0, "", 0, ""]
    
    matchedNode_df = pd.DataFrame(np.nan, index=range(len(dfSubset)), columns=list(["trip_id", "matchedLat", "matchedLon"]))
    matchedNode_df.iloc[0] = [0, 0, 0]
    
    trips_df_columns = dfSubset.columns.union(ways_df.columns.union(matchedNode_df.columns))
    trips_df = pd.DataFrame(np.nan, index=range(len(dfSubset)), columns=trips_df_columns)

    # For Snapping the road segments
    roadSnap_df = pd.DataFrame([list([""])], index=range(len(dfSubset)-1), columns=["roadSnap"])

    #ID of the trip being processed
    trip_id = dfSubset['trip_id'][0]
    
    #Road name of the previous trippoint
    previousRoadName = u''
    
    for j in range(len(dfSubset)):
        tripPoint = dfSubset.iloc[[j]]
        gpsLat = tripPoint['latitude'][j]
        gpsLon = tripPoint['longitude'][j]
            
        ways_df = nearestNeighbourWay(tripPoint, con, trip_id)
        
        #For the starting trippoint j==0
        if j == 0:
            selectedWay_df = ways_df.iloc[[0]]
            
            road = selectedWay_df['way'][0][11:len(selectedWay_df['way'][0])-1].split(',')
            matchingPointData = getMatchingPointData(road, gpsLat, gpsLon)
            matchedCoordinates = matchingPointData[0]
            matchedNode_df.iloc[0] = [trip_id, matchedCoordinates[0], matchedCoordinates[1]]
            
            # After processing trippoint, setting it as previous road name for processing the upcoming trippoint                
            previousRoadName = ways_df['name'][0]
        elif len(ways_df) < 1:
            matchedNode_df.iloc[0] = [trip_id, 0, 0]
            
            ways_df = pd.DataFrame(np.nan, index=[0], columns=list(["trip_id", "osm_id", "name", "mylinedistance", "way"]))
            ways_df.iloc[0] = [trip_id, 0, "", 0, ""]
            
        elif len(ways_df) > 0:
            shortestPathList = [1000]*len(ways_df)
            
            if sum(ways_df['name'].isin([previousRoadName])) > 0:
                bestIndex = list(ways_df['name']).index(previousRoadName)

                roadString = ways_df['way'][bestIndex][11:len(ways_df['way'][bestIndex])-1]
                road = roadString.split(',')
                matchingPointData = getMatchingPointData(road, gpsLat, gpsLon)
                matchedCoordinates = matchingPointData[0]
                matchedNode_df.loc[j] = [trip_id, matchedCoordinates[0], matchedCoordinates[1]]                
                
                shortestPathList[0] = getShortestPathResult(con, j, matchingPointData, matchedNode_df)
                
                roadSnap_df.append(shortestPathList[0])
                roadSnap_df.loc[j-1] = roadString
                
                # After processing trippoint, setting it as previous road name for processing the upcoming trippoint                
                previousRoadName = ways_df['name'][bestIndex]
            else:
                #SwitchArray is used to classify btw tricky and non-tricky points        
                switchArray = getSwitchArray(ways_df)
                
                #Variables to decide actual road for tricky trippoints
                distanceCost = [1000]*len(ways_df)
                pathCost = [1000]*len(ways_df)
                nearestWay = [1000]*len(ways_df)
                wayEstimationValue = [1000]*len(ways_df)
        
                for q in range(sum(switchArray)):
                    selectedWay_df = ways_df.iloc[[q]]
                    
                    road = selectedWay_df['way'][q][11:len(selectedWay_df['way'][q])-1].split(',')
                    matchingPointData = getMatchingPointData(road, gpsLat, gpsLon)

                    shortestPathList[q] = getShortestPathResult(con, j, matchingPointData, matchedNode_df)
                    
                    matchingPointDistance = matchingPointData[1]
                    shortestPathCost = shortestPathList[q]['cost'].sum()
                    nearestWayDistance = ways_df.iloc[q]['mylinedistance']
                    
                    # FINALLY, the Decision Value can be stated as w = x + y + z
                    distanceCost[q] = matchingPointDistance
                    pathCost[q] = shortestPathCost
                    nearestWay[q] = nearestWayDistance
                    wayEstimationValue[q] = matchingPointDistance + shortestPathCost + nearestWayDistance
                #END of FOR Loop
                
                # Getting the index of the way with least cost
                bestIndex = wayEstimationValue.index(min(wayEstimationValue))
                
                # Extracting the road segment
                roadString = ways_df['way'][bestIndex][11:len(ways_df['way'][bestIndex])-1]
                road = roadString.split(',')
                
                # Getting the matched points on the road segment
                matchingPointData = getMatchingPointData(road, gpsLat, gpsLon)
                matchedCoordinates = matchingPointData[0]
                matchedNode_df.loc[j] = [trip_id, matchedCoordinates[0], matchedCoordinates[1]]
                
                # Snapping the road segments 
                roadSnap_df.loc[j-1] = shortestPathList[bestIndex]['st_astext']

                # After processing trippoint, setting it as previous road name for processing the upcoming trippoint                
                previousRoadName = ways_df['name'][bestIndex]
            #END of inner ELSE
        #END of last ELSE
        
        trips_df.loc[[j]] = pd.merge(pd.merge(tripPoint, selectedWay_df, how='outer', on=['trip_id']), matchedNode_df, how='outer', on=['trip_id'])
    #END of for loop        
    
    map_osm = viewMatchedPointsOnMap(map_osm, trips_df)
    
    map_osm = viewRoadSnapOnMap(map_osm, roadSnap_df, trips_df)
    
    return([trips_df, map_osm])
#END of processTrip()

#### viewMatchedPointsOnLeaflet(): View gps points and matched points on leaflet
def viewMatchedPointsOnMap(m, trips_df):
    gpxData = [trips_df['latitude'], trips_df['longitude']]
    matchedData = [trips_df['matchedLat'], trips_df['matchedLon']]
    
    for k in range(len(trips_df)):
        folium.CircleMarker(location=[gpxData[0][k], gpxData[1][k]], radius=4, color='#31a354', fill_color='#31a354').add_to(m)
        folium.CircleMarker(location=[matchedData[0][k], matchedData[1][k]], radius=5, color='red', fill_color='red').add_to(m)
    #END of for loop
    
    return(m)
#END of viewMatchedPointsOnLeaflet()
    
#### viewRoadSnapOnMap(): View road snap for a trip on leaflet
def viewRoadSnapOnMap(map_osm, roadSnap_df, trips_df):
    if(len(roadSnap_df) > 0):
        for w in range(len(roadSnap_df)):
            roadSegment = roadSnap_df['roadSnap'][w][11:len(roadSnap_df['roadSnap'][w])-1].split(',')

            lonVector = [0]*len(roadSegment)
            latVector = [0]*len(roadSegment)            
            
            for u in range(len(roadSegment)):
                latlon = roadSegment[u].split(" ")
                lonVector[u] = latlon[0]
                latVector[u] = latlon[1]
            #END of FOR loop
            
            for k in range(len(latVector)):
                folium.CircleMarker(location=[latVector[k], lonVector[k]], radius=2.5, color='#blue', fill_color='#blue').add_to(map_osm)
                #folium.PolyLine([latVector[k], lonVector[k]], color="black", weight=2, opacity=1).add_to(map_osm)
            #END of for loop            
        # END of outer FOR LOOP
            
    return(map_osm)
#END of viewRoadSnapOnMap()

#### getShortestPathResult(): Execute Dijkstra to find shortest path between 2 osm nodes
def getShortestPathResult(con, j, matchingPointData, matchedNode_df):
    matchedCoordinates = matchingPointData[0]
    
    sourceLon = round(matchedNode_df.iloc[j-1][2], 5)
    sourceLat = round(matchedNode_df.iloc[j-1][1], 5)
    targetLon = round(matchedCoordinates[1], 5)
    targetLat = round(matchedCoordinates[0], 5)
            
    sourceNodeQuery = ("SELECT source FROM hh_2po_4pgr ORDER BY geom_way <-> "+
                        "ST_SetSRID(ST_Makepoint(%f,%f), 4326) LIMIT 1;") % (sourceLon, sourceLat)
    targetNodeQuery = ("SELECT target FROM hh_2po_4pgr ORDER BY geom_way <-> "+
                        "ST_SetSRID(ST_Makepoint(%f,%f), 4326) LIMIT 1;") % (targetLon, targetLat)
                                    
    sourceNode = pd.read_sql_query(sourceNodeQuery,con=con).iloc[0][0]
    targetNode = pd.read_sql_query(targetNodeQuery,con=con).iloc[0][0]
    
    #boundingBox = getBoundingBox(targetLat, targetLon, 5.0)
    
    #mnLon = boundingBox[0]
    #mnLat = boundingBox[1] 
    #mxLon = boundingBox[2]
    #mxLat = boundingBox[3]
    
    if sourceNode != targetNode :            
        shortestPathQuery = ("SELECT seq, id1 AS node, id2 AS edge, di.cost, ST_AsText(ST_Transform(geom_way, 4326))"+ 
                            " FROM pgr_dijkstra('SELECT id, source, target, st_length(geom_way) as cost FROM hh_2po_4pgr WHERE "+
                            " public.hh_2po_4pgr.geom_way && ST_MakeEnvelope(13.724327,45.606592,13.848267,45.70534, 4326)',"
                            "%d, %d, false, false ) as di JOIN hh_2po_4pgr ON di.id2 = hh_2po_4pgr.id") % (sourceNode,targetNode)
        #print(shortestPathQuery)
        return(pd.read_sql_query(shortestPathQuery,con=con).iloc[0])
    else:
        return(pd.DataFrame([-1000], index=[0], columns=["cost"]))
#END of getShortestPathResult()

#### nearestNeighbourWay(): Get nearest way from PostGIS using the getBoundingBox() function
def nearestNeighbourWay(tripPoint, con, trip_id):    
    radLat = tripPoint['latitude']
    radLon = tripPoint['longitude']
    boundingBox = getBoundingBox(radLat, radLon, 1.2)
    
    minLon = boundingBox[0]
    minLat = boundingBox[1] 
    maxLon = boundingBox[2]
    maxLat = boundingBox[3]
    
    spatialQuery = ("with distance_query as(SELECT DISTINCT ON (way)"
    +" osm_id, name, ST_Distance(ST_Transform(ST_GeomFromText('POINT(%f %f)',4326),3857),"
    +" ST_Transform(way, 3857)) AS myLineDistance, ST_AsText(ST_Transform(way, 4326)) as way"
    +" FROM public.planet_osm_line"
    +" WHERE public.planet_osm_line.way && ST_Transform("
    +" ST_MakeEnvelope(%f,%f,%f,%f, 4326), 3857)"
    +" AND ST_Distance(ST_Transform(ST_GeomFromText('POINT(%f %f)',4326),3857),"
    +" ST_Transform(way, 3857)) < 10.0 AND osm_id > 0)"
    +" SELECT *FROM distance_query" 
    +" ORDER BY myLineDistance LIMIT 3;"
    ) % (radLon, radLat, minLon, minLat, maxLon, maxLat, radLon, radLat) 

    #way: Execute the query and get the results in a simple df with geometry as text
    ways_table = pd.read_sql_query(spatialQuery,con=con)
    
    #Keep trip_id in the ways dataframe, to keep track and later on concat results
    ways_table['trip_id'] = [trip_id]*len(ways_table)
  
    return(ways_table)  
#END of nearestNeighbourWay()

#### getBoundingBox(): Get a bounding box around lat/lon for some distance
def getBoundingBox(radLat, radLon, distance):
    eRadius = 6371.01
    
    pRadius = eRadius*math.cos(radLat)
    
    minLat = radLat - distance/eRadius
    maxLat = radLat + distance/eRadius
    minLon = radLon - distance/pRadius
    maxLon = radLon + distance/pRadius
    
    boundingBox = list([minLon, minLat, maxLon, maxLat])
    
    return(boundingBox)
#END of getBoundingBox()

#### Calculate the great circle distance between two points on the earth
def haversine(lon1, lat1, lon2, lat2):
    # convert decimal degrees to radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])

    # haversine formula 
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    r = 6371 # Radius of earth in kilometers. Use 3956 for miles
    return c * r
#END of haversine
  
#### getMatchingPointData(): Get the matching point to the GPS point on the road
def getMatchingPointData(road, gpsLat, gpsLon):
    distances = []
    
    for node in road:
        latlon = node.split(" ")
        lat = float(latlon[1])
        lon = float(latlon[0])
        
        d = haversine(lon, lat, gpsLon, gpsLat)
        distances.append(d)
    #END of for loop
        
    minimumNodeIndex = distances.index(min(distances))
    matchedNode = road[minimumNodeIndex]
    latlon = matchedNode.split(" ")
    matchedLat = float(latlon[1])
    matchedLon = float(latlon[0])
  
    returnList = []
    returnList = [[matchedLat, matchedLon], distances[minimumNodeIndex]]
  
    return(returnList)
#END of getMatchingPointData()
        
#### getSwitchArray(): Array used later to classify tricky and non-tricky trippoints
def getSwitchArray(ways_df):
    wayDistances = ways_df['mylinedistance']
    meanValue = wayDistances.mean()
    returnArray = [0]*len(ways_df)
    
    if len(wayDistances) > 1:
        returnArray[0] = 1
        for x in range(len(wayDistances)):
            #Check to take care of last x value
            if x < len(wayDistances)-1:
                diff = wayDistances[x+1] - wayDistances[0]
                #Checks to classify tricky and non-tricky trippoints    
                if diff == 0:
                    break
                elif diff < meanValue:
                    returnArray[x+1] = 1
                else:
                    break
        #END of FOR loop
    else:
        returnArray[0] = 1
    #END of OUTER ELSE
    
    return(returnArray)
#END of getSwitchArray()
